master/buildbot/process/build.py

```diff
@@ -542,30 +529,38 @@ def buildFinished(self, text, results):
         If 'results' is SUCCESS or WARNINGS, we will permit any dependant
         builds to start. If it is 'FAILURE', those builds will be
         abandoned."""
-        self.stopBuildConsumer.stopConsuming()
-        self.finished = True
-        if self.conn:
-            self.subs.unsubscribe()
-            self.subs = None
-            self.conn = None
-        log.msg(" %s: build finished" % self)
-        self.results = worst_status(self.results, results)
-        self.build_status.setText(text)
-        self.build_status.setResults(self.results)
-        self.build_status.buildFinished()
-        eventually(self.releaseLocks)
-        self.deferred.callback(self)
-        self.deferred = None
+        try:
+            self.stopBuildConsumer.stopConsuming()
+            self.finished = True
+            if self.conn:
+                self.subs.unsubscribe()
+                self.subs = None
+                self.conn = None
+            log.msg(" %s: build finished" % self)
+            self.results = worst_status(self.results, results)
+            self.build_status.setText(text)
+            self.build_status.setResults(self.results)
+            self.build_status.buildFinished()
+            eventually(self.releaseLocks)
+            metrics.MetricCountEvent.log('active_builds', -1)
+
+            yield self.master.data.updates.setBuildStateString(self.buildid,
```

Comment:
This is just the stuff that happened after wiating for the defered above in `Build.startBulid` and `Builder._maybeStartBuild`.